18447 Fix opening of preview areas on clicking preview button

myabc opened the pull request:
https://community.openproject.org/work_packages/18447

myabc committed:
Rework broken preview code as hasPreview directive

Bug # 18447

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Rename PreviewsHelper arg to reflect actual usage

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix preview_link usages to use link not form id

Combined with previous commit (c95f5c54), this should ensure that we
now create a valid DOM, i.e. without duplicate IDs.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://fix-18447-preview-button-open-preview.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://fix-18447-preview-button-open-preview.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc commented:
https://community.openproject.org/work_packages/18447

myabc commented:
@floriank Not sure if you still wanted to take a look at this. As-is this _works for me_, but Cucumber features are currently failing and I've yet to find time to debug them.

I did a minimal port of jQuery code to Angular. Further scope for refactoring:
- switching to Angular `$http`, `$anchorScroll` services (not tested though): https://gist.github.com/myabc/7506ad8b27eb35a6a019
- creating a `PreviewService` or such like.
- removing the hard-coded target element (`#preview`) and making as second `previewArea` directive that registers itself as being interested in previews for a particular form (not sure how best to handle intra-directive communication here).

floriank commented:
I can verify the functionality - it seems to work fine. 

I'll have a look at the cucumber tests.

myabc commented:
@floriank thanks!

floriank commented:
@myabc @ulferts It seems that the PR actually broke the issue update view, as the angular directive `has-preview` is not properly compiled due to the fact that the element is created after the dom has been loaded. 

I'll be fixing it. Dunno how yet - I am open for suggestions, though.

Commit:
fix the issue edit view

The problem was the edit view which is loaded via ajax and attached to
the DOM. The angular directive has-preview is not run for the newly
attached edit view. The directive was manually compiled to resolve the
issue.

Commit:
fix a minification issue

floriank commented:
This PR will also solve https://community.openproject.org/work_packages/18451

ulferts commented on app/assets/javascripts/application.js.erb:
```diff
@@ -1213,23 +1213,3 @@ var SubmitConfirm = (function($) {
-        return false;
-      });
-    });
-})(jQuery);
```
:+1: for removing code from the application.js
